### PR TITLE
eratta サモン・ソーサレス

### PR DESCRIPTION
**Pull request by purerosefallen**
Rework of https://github.com/Fluorohydride/ygopro-scripts/pull/2458

The above PR does not accept changes from maintainers, so I opened a new one.

### CODE REVIEW

File: c61665245.lua [hunk 1]
```diff
@@ -10,117 +10,72 @@ function c61665245.initial_effect(c)
...
-function c61665245.sptg1(e,tp,eg,ep,ev,re,r,rp,chk)
+function c61665245.spfilter2(c,e,tp,rac)
+	return c:IsRace(rac) and c:IsCanBeSpecialSummoned(e,0,tp,false,false)
+end
+function c61665245.sptg(e,tp,eg,ep,ev,re,r,rp,chk)
 	local zone=e:GetHandler():GetLinkedZone(1-tp)
 	if chk==0 then return Duel.GetLocationCount(1-tp,LOCATION_MZONE,tp,LOCATION_REASON_TOFIELD,zone)>0
 		and Duel.IsExistingMatchingCard(c61665245.spfilter1,tp,LOCATION_HAND,0,1,nil,e,tp,zone) end
-	Duel.SetOperationInfo(0,CATEGORY_SPECIAL_SUMMON,nil,1,tp,LOCATION_HAND)
+	Duel.SetOperationInfo(0,CATEGORY_SPECIAL_SUMMON,nil,2,tp,LOCATION_HAND+LOCATION_DECK)
```
Comment (mercury233) [inline c61665245.lua:35]:
```suggestion
	Duel.SetOperationInfo(0,CATEGORY_SPECIAL_SUMMON,nil,1,tp,LOCATION_HAND+LOCATION_DECK)
```

File: c61665245.lua [hunk 2]
```diff
@@ -10,117 +10,72 @@ function c61665245.initial_effect(c)
...
 end
-function c61665245.spop1(e,tp,eg,ep,ev,re,r,rp)
+function c61665245.spop(e,tp,eg,ep,ev,re,r,rp)
+	local c=e:GetHandler()
 	local zone=e:GetHandler():GetLinkedZone(1-tp)
```
Comment (mercury233) [inline c61665245.lua:39]:
```suggestion
	local zone=c:GetLinkedZone(1-tp)
```

File: c61665245.lua [hunk 3]
```diff
@@ -10,117 +10,72 @@ function c61665245.initial_effect(c)
...
-	local tc=Duel.GetFirstTarget()
-	if not c:IsRelateToEffect(e) or not tc:IsRelateToEffect(e) or tc:IsFacedown() then return end
-	local zone={}
-	local flag={}
-	for p=0,1 do
-		zone[p]=c:GetLinkedZone(p)&0xff
-		local _,flag_tmp=Duel.GetLocationCount(p,LOCATION_MZONE,tp,LOCATION_REASON_TOFIELD,zone[p])
-		flag[p]=(~flag_tmp)&0x7f
+	local tc=Duel.SelectMatchingCard(tp,c61665245.spfilter1,tp,LOCATION_HAND,0,1,1,nil,e,tp,zone):GetFirst()
+	local function makeDisable(dc)
```
Comment (mercury233) [inline c61665245.lua:43]: should be separate

File: c61665245.lua [hunk 4]
```diff
@@ -10,117 +10,72 @@ function c61665245.initial_effect(c)
...
+				makeDisable(sc)
+				local e3=Effect.CreateEffect(c)
+				e3:SetType(EFFECT_TYPE_FIELD)
+				e3:SetCode(EFFECT_CANNOT_SPECIAL_SUMMON)
+				e3:SetProperty(EFFECT_FLAG_PLAYER_TARGET)
+				e3:SetTargetRange(1,0)
+				e3:SetTarget(c61665245.splimit)
+				e3:SetLabel(sc:GetOriginalRace())
+				e3:SetReset(RESET_PHASE+PHASE_END)
+				Duel.RegisterEffect(e3,tp)
```
Comment (mercury233) [inline c61665245.lua:74]: summon complete?

**Review by mercury233 — CHANGES_REQUESTED** [review]

File: c61665245.lua (hunk 4 above)
Comment (purerosefallen) [reply c61665245.lua:74]: summon complete is at the top function part

File: c61665245.lua (hunk 3 above)
Comment (purerosefallen) [reply c61665245.lua:43]: it's should not separate, because the function referenced `c` which is this card itself.

File: c61665245.lua [hunk 5]
```diff
@@ -10,117 +10,72 @@ function c61665245.initial_effect(c)
...
+		e1:SetCode(EFFECT_DISABLE)
+		e1:SetReset(RESET_EVENT+RESETS_STANDARD)
+		dc:RegisterEffect(e1)
+		local e2=Effect.CreateEffect(c)
+		e2:SetType(EFFECT_TYPE_SINGLE)
+		e2:SetCode(EFFECT_DISABLE_EFFECT)
+		e2:SetValue(RESET_TURN_SET)
+		e2:SetReset(RESET_EVENT+RESETS_STANDARD)
+		dc:RegisterEffect(e2)
+		Duel.SpecialSummonComplete()
```
Comment (purerosefallen) [inline c61665245.lua:55]: @mercury233 right here.

Comment (mercury233) [reply c61665245.lua:55]: don't put it into a function whose name has no relation

**Comment by salix5** [comment]
https://github.com/Fluorohydride/ygopro-scripts/wiki/Category
CATEGORY_DECKDES
Effects that include "Send a card from the Deck to the GY" or "Special Summon from the Deck".
Usage: 14558127 灰流うらら

**Comment by purerosefallen** [comment]
@salix5 
changed.

Could we change it into something like `CATEGORY_LEAVE_DECK_URARA` to avoid misleading names?
In fact at the time when Urara came out, this also became another purposal.

File: c61665245.lua (hunk 5 above)
Comment (purerosefallen) [reply c61665245.lua:55]: how about rename that function as `disableAndCompleteSummon(c)`?

**Pull request merged**